### ISSUE TITLE
fastboot.isFastBoot should be false in fallback mode

### ISSUE THREAD

**Issue by Redsandro** [ember-fastboot/ember-cli-fastboot#291]
I might be wrong because my code sandbox is a bit messy, but when falling back to non-fastboot mode (e.g. on error), it seems that `fastboot.isFastBoot` is still `true`.

This is undesirable, e.g. when using `fastboot.isFastboot` for switching between `fastboot.request.headers` and `window.navigator`. In fallback, the wrong code gets executed because it thinks it's still running in Fastboot.

**Comment by danmcclain** [comment]
What do you mean by fallback mode? Are you referring to when an error happens and the non-FastBoot index.html is served?

**Comment by rwjblue** [comment]
Can you provide a demo repo and instructions to reproduce what you are experiencing?

**Comment by Redsandro** [comment]
> Are you referring to when an error happens and the non-FastBoot index.html is served?

Exactly. If the entire app works but a certain route accidentally has a (fastboot-only) error, fastboot falls back so the app still runs. But `fastboot.isFastboot` is still true.

Or perhaps I am using it wrong? Is there a different flag to check if this is a fastboot-render or normal (client) render?

**Comment by Redsandro** [comment]
@rwjblue 
> Can you provide a demo repo and instructions to reproduce what you are experiencing?

Let me try and just show some code first. Perhaps you can already notice something without me getting my hands dirty on a repo. :wink: 

Because if your initial thought is not: _"Yes, `isFastboot` is always `true` when running a fastboot server, even in fallback mode"_, then I probably made a mistake myself.

```
import Ember from 'ember';

export default Ember.Controller.extend({
	fastboot : Ember.inject.service(),
	init : function() {
		this._super(...arguments);

		Ember.run.schedule("afterRender", this, function() {
			if (!this.get('fastboot').isFastboot)
				this.send("onLayoutComplete");
		});
	},
	actions : {
		onLayoutComplete() {
			Ember.Logger.info("This should only be executed in fallback mode");
			// Some random jQuery
			$('body').css({backgroundColor: 'orange'});
		}
	}
});
```

Fastboot still fails because `$` is (obviously) not defined. I would expect the `isFastboot` switch to prevent this.

**Comment by danmcclain** [comment]
The FastBoot service is destroyed after visit completes, so it is false in the run later. Try wrapping the run later in the FastBoot check, rather than checking inside of the run loop

**Comment by Redsandro** [comment]
In that case I'm lacking some Ember smarts. I thought `Ember.run.schedule("afterRender", ...)` was as late as it would go.

I know it's probably not too interesting to deal with people who aren't fully updated with the Ember spec, but would you be willing to give me a suggestion for how/where to put a post-fastboot run?

**Comment by danmcclain** [comment]
You effectively want this:
```
import Ember from 'ember';

export default Ember.Controller.extend({
    fastboot : Ember.inject.service(),
    init : function() {
        this._super(...arguments);

        if (!this.get('fastboot').isFastboot) {
          Ember.run.schedule("afterRender", this, function() {
            this.send("onLayoutComplete");
          });
       }
    },
    actions : {
        onLayoutComplete() {
            Ember.Logger.info("This should only be in the browser");
            // Some random jQuery
            $('body').css({backgroundColor: 'orange'});
        }
    }
});
```

The ember app re-evaluates on the browser side (I think this is what you are intending with "fallback" mode)

**Comment by Redsandro** [comment]
@danmcclain 
> The FastBoot service is destroyed after visit completes, so it is false in the run later. 

This does not seem to be true. This will execute `onRenderComplete` twice. Once in the fastboot run, and once on the client side.

```
if (!this.get('fastboot').isFastboot) {
	Ember.run.schedule('afterRender', this, function() {
		this.send("onRenderComplete");
	});
}
```

**Comment by danmcclain** [comment]
Ah, capitalize the b in `isFastBoot` and the code will stop running server side. Sorry for the typo in my example. Fixing that will do what you intend. The if has to be outside the run later, or you have to capture the value before using it in the run later

**Comment by Redsandro** [comment]
Oh! Thank you, I didn't even notice I copied the 'typo'. I feel a bit silly for not noticing myself.

So in FastBoot mode, some events will fire double on first pageview? Or all events? Is there documentation about what events are fired on which side? (fastboot/server or client)

I tried to look up the list of events, and the fastboot influence on them, but I couldn't find it. I apologize if I inadvertently missed an obvious piece of information.

**Comment by Redsandro** [comment]
@danmcclain it appears the opposite is true.

`this.get('fastboot').isFastBoot` will always be `true` even if ember is started in non-fastboot mode (with `ember serve`) in stead of `ember fastboot`.

**Comment by danmcclain** [comment]
`isFastBoot` is only true when running serverside, within node, so here is the truth table:

 * ember fastboot
   * Server side: true ✅
   * Client side: false ❌
 * ember serve
   * Client side: false ❌

Can you push an example exhibiting when it is `true` in the browser?

**Comment by danmcclain** [comment]
@Redsandro Are you still seeing this issue? Can you provide a small sample repo that exhibits the behavior?

**Comment by Redsandro** [comment]
@danmcclain Sorry for the late reply. I'm quite(*) sure that last boolean in your truth table is wrong and I was meaning to get back to you and prove it using code. However december is being very december if you know what I mean, so I'm hoping you're willing to leave this open a little bit longer for me to respond.

If it's helping, in your truth table (I love that term) you use graphics to reflect the value of the booleans. In my truth table, I was using graphics to illustrate whether or not the boolean has the value I'd expect it to be.

*) Was. Was quite sure. If I was an innocent bystander I'd bet on you. But I have to protect my ego somewhat. So I'd say that there is an 80% chance I'm wrong about this. I'll have to isolate the code and see if the fail of the boolean was an error of my own making.

**Comment by Redsandro** [comment]
__Some controller__

```javascript
import Ember from 'ember';

export default Ember.Controller.extend({
	fastboot	: Ember.inject.service(),
	init : function() {
		this._super(...arguments);

		console.log('FASTBOOT: This should show up both on server and client.');

		if (this.get('fastboot').isFastBoot) {
			console.log('FASTBOOT: This should only show up on the server. Not the client.')
		}
    }
});
```

__Ember server, as expected:__

```cmd
Installing FastBoot npm dependencies
DEBUG: -------------------------------
DEBUG: Ember      : 2.7.0
DEBUG: Ember Data : 2.7.0
DEBUG: -------------------------------
Ember FastBoot running at http://[::]:3000
Preparing to transition from '' to 'test/controller'
FASTBOOT: This should show up both on server and client.
FASTBOOT: This should only show up on the server. Not the client.
```

__Developer console, not expecting second message:__

![image](https://cloud.githubusercontent.com/assets/1702193/21158011/e1519a12-c17b-11e6-8439-b1f8f9fdde92.png)

**Comment by Redsandro** [comment]
__Workaround__

So basically I want to do this, but it doesn't work:

```javascript
import Ember from 'ember';

export default Ember.Controller.extend({
  fastboot  : Ember.inject.service(),
    init() {
    this._super(...arguments);

    if (!this.get('fastboot').isFastBoot) {
      Ember.run.schedule('afterRender', this, function() {
        this.send("onRenderComplete");
      });
    }
  },
  actions    : {
    onRenderComplete() {
      // I want to do something with jquery, but this code never executes.
    }
  }
});
```

So to workaround this, I use the following:

```javascript
import Ember from 'ember';

export default Ember.Controller.extend({
  fastboot  : Ember.inject.service(),
    init() {
    this._super(...arguments);

    Ember.run.schedule('afterRender', this, function() {
      this.send("onRenderComplete");
    });
  },
  actions    : {
    onRenderComplete() {
      if (window.$) {
          // Do something with jQuery
      }
      else {
        Ember.Logger.debug("DOM not available.");
      }
    }
  }
});
```

__Truth table__

| Server                  | observed | expected |
|-------------------------|----------|----------|
| `ember fastboot` server | `true`   | `true`   |
| `ember fastboot` client | `true`   | `false`  |
| `ember serve` client    | `false`  | `false`  |

**Comment by danmcclain** [comment]
The issue is you need to call `this.get('fastboot.isFastBoot')`, when you call `this.get('fastboot').isFastBoot`, it is returning the computed property, not the final value.

**Comment by jasonmit** [comment]
@danmcclain we can make this a normal prop, which would avoid future trolling.  Want me to PR?

**Comment by danmcclain** [comment]
@jasonmit that would be awesome

**Comment by Redsandro** [comment]
I was going to quote bad examples I used from some docs or readme in order to propose clearing this up there, because I sincerely did copy this formatting with the best of intentions, but I notice the docs are actually quite clear (and correct). I've been watching this project for a while so I'm not sure how old the docs I used were, but with the current examples I would not have messed up.

Although this mistake could have been observed from my initial example code in this issue, I feel a bit silly. My apologies for wasting your time.

---

(typed 4 hours ago, tab still open)
